firmware/library/L2_HAL/memory/sddriver.hpp

```diff
@@ -0,0 +1,1169 @@
+  //              blocks of data from an SD Card
+  virtual uint8_t DeleteBlock(uint32_t start, uint32_t end) = 0;
+
+  // @description     This function performs a CRC-7 Add Operation
+  // @parameter       (uint8_t) CRC
+  //                                          The CRC to add upon
+  // @parameter       (uint8_t) message_byte
+  //                                          The byte to use when
+  //                                          performing the add operation.
+  // @returns         (uint8_t) newCRC
+  //                                          The new CRC-7 value resulting
+  //                                          from the add operation.
+  virtual uint8_t Crc7Add(uint8_t CRC, uint8_t message_byte) = 0;
+
+  // @description     This function acquires the CRC-7 code of a given
+  //                  message with the given length
+  // @parameter       (uint8_t*) message
+  //                                          A pointer to an array housing
+  //                                          the message.
+  // @parameter       (uint8_t) length        The length of the message
+  // @returns         (uint8_t) crc
+  virtual uint8_t GetCrc7(uint8_t message[], uint8_t length) = 0;
+
+  // @description     This function acquires the CRC-16 code of a given
+  //                  message with the given length
+  // @parameter       (uint8_t*) message
+  //                                          A pointer to an array housing
+  //                                          the message.
+  // @parameter       (uint16_t) length        The length of the message
+  // @returns         (uint16_t) newCRC  The new CRC-16 value as a
+  //                                          result from the calculation.
+  virtual uint16_t GetCrc16(uint8_t message[], uint16_t length) = 0;
+};
+
+class Sd : public SdInterface
+{
+ public:
+  // This value was found through experimentation. The suggested value (found
+  // via research) is to wait at least 1-8 bytes after sending a command's CRC
+  // before expecting a response from the card.
+  // See https://luckyresistor.me/cat-protector/software/sdcard-2/
+  static constexpr uint8_t kBusTimeout = 250;
+
+  struct DebugSdCard
+  {
+  };
+
+  // Enforcing block-size cross-compatibility
+  static constexpr uint16_t kBlockSize = 512;
+  static constexpr CrcTableConfig<uint8_t> kCrcTable8 =
+      GenerateCrc7Table<uint8_t>();
+  static constexpr CrcTableConfig<uint16_t> kCrcTable16 = GenerateCrc16Table();
+
+  constexpr Sd(uint8_t port = 1, uint8_t pin = 25)
+      : ssp_interface_(&ssp_),
+        ssp_(Ssp::Peripheral::kSsp0),
+        chip_select_(&chip_select_pin_),
+        chip_select_external_(&chip_select_external_pin_),
+        chip_select_pin_(port, pin),
+        chip_select_external_pin_(5, 3)
+  {
+  }
+
+  // This version of the constructor will provide you with an external pin and
+  // port that is broken out through the board's GPIO, allowing you to probe
+  // the chip select manually.
```

Comment:
Put an example of how to trigger this constructor in the comment here.

Reply:
Done in my staging branch.